components/function-controller/design/subresources_templates_proposal.md

```diff
@@ -0,0 +1,84 @@
+# Function Subresources Templating for Serverless v1alpha2
+
+## Summery
+
+The current Serverless API allows for limited ability to configure the generated functions subresources (deployment, build job). Additionally, it doesn't allow for separate configuration for build and functions resources except for memory/cpu resources.
+
+This proposes changes to the current v1alpha2 Serverless API to support templating of function sub resources _without_ the need to rollout a new API version.
+
+## Motivation
+
+Give Serverless users the ability to:
+- Separate configuration for function deployments and build jobs.
+- Configure volume mounts for function subresources
+- Configure sidecars for function deployments (?)
+
+### Goals
+
+- Add more flexibility to the Serverless API.
+- Provide a rollout plan to implement this functionality in small increments and avoid the need to rollout a new API version
+
+### Non-Goals
+- Define a new Spec API version
+
+## Proposal
+
+Use a modified version of [PodTemplateSpec](https://github.com/kubernetes/kubernetes/blob/64ed9145452d2d1d324d2437566f1ea1ce76f226/pkg/apis/core/types.go#L3443) as a base for function build/deployment templates. 
+
+A modified version is needed because we need to protect certain parts of the Pod spec (commands and args for the main container in the pod for example).
+
+The proposal doesn't include removing or moving any existing Spec fields and only adds non-breaking changes to the Spec. This ensure that there is no need to rollout a new API version and ensure backward compatibility with released v1alpha2 Spec/API.
+
+## Design details
+
+Currently, we expose the following _relevant_ high level Spec fields:
+- `Env`
+- `ResourceConfiguration`
+- `ScaleConfig`
+- `Replicas`
+- `Template` (Labels and annotations, currently not applied)
+
+The proposed functionality can be implemented by adding the two following fields to the spec:
+- `RuntimeTemplate`
+- `BuildTemplate`
+
+*Note:* We should decide if we want to the implement this directly under the function Spec or grouped in a Spec field, e.g. `spec.Templates`.
+
+Both fields will be of the modified version of the PodTemplateSpec.
+
+Both fields will be optional. They should be defaulted based on the values of the high level Spec fields.
+
+### Defaulting
+
+The high level Spec fields should be defaulted by the _defaulting webhook_.
+
+If the template fields are not set, they should be defaulted based on the high level Spec by the _function controller_. 
```

Comment:
Why should we default template values?

Reply:
- I _think_ it would be easier to implement the reconciliation code this way. The controller will initially fill in the template based on the high level spec, defaults and/or user overrides, then the rest of the logic will only use the template spec during reconcile. This VS having reconciliation code that handles two specs based on what's defined and what's not.

- It will allow the user to _see_ the full spec in the CR. This allows users (and us) to have more visibility on the internal spec and ability to fine tune it as needed.